File: packages/expo-ui/ios/TextInput/TextInputMask/ForkInputMask/Source/InputMask/InputMask/Classes/Model/State/FreeState.swift
//
// Project «InputMask»
// Created by Jeorge Taflanidi
//


import Foundation


/**
 ### FreeState
 
 Represents "free" characters outside square and curly brackets.
 
 Accepts every character but does not put it into the result string, unless the character equals the one from the mask
 format. If it's not, inserts the symbol from the mask format into the result.
 
 Always returns `nil` as an extracted value, does not affect the resulting value.
 */
class FreeState: MaskState {
    
    let ownCharacter: Character
    
    override func accept(character char: Character) -> Next? {
        if self.ownCharacter == char {
            return Next(
                state: self.nextState(),
                insert: char,
                pass: true,
                value: nil
            )
        } else {
            return Next(
                state: self.nextState(),
                insert: self.ownCharacter,
                pass: false,
                value: nil
            )
        }
    }
    
    override func autocomplete() -> Next? {
        return Next(
            state: self.nextState(),
            insert: self.ownCharacter,
            pass: false,
            value: nil
        )
    }
    
    /**
     Constructor.
     
     - parameter child: next ``State``
     - parameter ownCharacter: represented "free" character
     
     - returns: Initialized ``FreeState`` instance.
     */
    init(
        child: MaskState,
        ownCharacter: Character
    ) {
        self.ownCharacter = ownCharacter
        super.init(child: child)
    }
    
    override var debugDescription: String {
        return "\(self.ownCharacter) -> " + (nil != self.child ? self.child!.debugDescription : "nil")
    }
    
}

File: packages/expo-ui/android/src/main/java/expo/modules/ui/textfield/model/state/FreeState.kt
package expo.modules.ui.textfield.model.state

import expo.modules.ui.textfield.model.Next
import expo.modules.ui.textfield.model.State

/**
 * ### FreeState
 *
 * Represents "free" characters outside square and curly brackets.
 *
 * Accepts every character but does not put it into the result string, unless the character equals
 * the one from the mask format. If it's not, inserts the symbol from the mask format into the
 * result.
 *
 * Always returns `null` as an extracted value, does not affect the resulting value.
 *
 * @author taflanidi
 */
class FreeState(child: State, val ownCharacter: Char) : State(child) {

    override fun accept(character: Char): Next? {
        return if (this.ownCharacter == character) {
            Next(
                this.nextState(),
                character,
                true,
                null
            )
        } else {
            Next(
                this.nextState(),
                this.ownCharacter,
                false,
                null
            )
        }
    }

    override fun autocomplete(): Next? {
        return Next(
            this.nextState(),
            this.ownCharacter,
            false,
            null
        )
    }

    override fun toString(): String {
        return "" + this.ownCharacter + " -> " + if (null == this.child) "null" else child.toString()
    }
}
